// @clerk/expo/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  getClerkInstance: () => import_singleton.getClerkInstance,
  isClerkAPIResponseError: () => import_errors.isClerkAPIResponseError,
  isClerkRuntimeError: () => import_errors.isClerkRuntimeError,
  isEmailLinkError: () => import_errors.isEmailLinkError,
  isKnownError: () => import_errors.isKnownError,
  isMetamaskError: () => import_errors.isMetamaskError
});
module.exports = __toCommonJS(index_exports);
var import_internal = require("@clerk/react/internal");
var import_errors = require("@clerk/react/errors");
var import_singleton = require("./provider/singleton");
__reExport(index_exports, require("./provider/ClerkProvider"), module.exports);
__reExport(index_exports, require("./hooks"), module.exports);
__reExport(index_exports, require("./components"), module.exports);
(0, import_internal.setErrorThrowerOptions)({ packageName: "@clerk/expo" });
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getClerkInstance,
  isClerkAPIResponseError,
  isClerkRuntimeError,
  isEmailLinkError,
  isKnownError,
  isMetamaskError,
  ...require("./provider/ClerkProvider"),
  ...require("./hooks"),
  ...require("./components")
});
