// @clerk/expo/dist/native/InlineUserProfileView.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InlineUserProfileView_exports = {};
__export(InlineUserProfileView_exports, {
  InlineUserProfileView: () => InlineUserProfileView
});
module.exports = __toCommonJS(InlineUserProfileView_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("@clerk/react");
var import_react2 = require("react");
var import_react_native = require("react-native");
var import_NativeClerkUserProfileView = __toESM(require("../specs/NativeClerkUserProfileView"));
var import_native_module = require("../utils/native-module");
function InlineUserProfileView({ isDismissable = false, style }) {
  const clerk = (0, import_react.useClerk)();
  const signOutTriggered = (0, import_react2.useRef)(false);
  const handleProfileEvent = (0, import_react2.useCallback)(
    async (event) => {
      var _a;
      const { type } = event.nativeEvent;
      if (type === "signedOut" && !signOutTriggered.current) {
        signOutTriggered.current = true;
        try {
          await ((_a = import_native_module.ClerkExpoModule) == null ? void 0 : _a.signOut());
        } catch (e) {
          if (false) {
            console.warn("[InlineUserProfileView] Native signOut error (may already be signed out):", e);
          }
        }
        if (clerk == null ? void 0 : clerk.signOut) {
          try {
            await clerk.signOut();
          } catch (err) {
            if (false) {
              console.warn("[InlineUserProfileView] JS SDK sign out error:", err);
            }
          }
        }
      }
    },
    [clerk]
  );
  if (!import_native_module.isNativeSupported || !import_NativeClerkUserProfileView.default) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_native.View, { style: [styles.container, style], children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_native.Text, { style: styles.text, children: !import_native_module.isNativeSupported ? "Native InlineUserProfileView is only available on iOS and Android" : 'Native InlineUserProfileView requires the @clerk/expo plugin. Add "@clerk/expo" to your app.json plugins array.' }) });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_NativeClerkUserProfileView.default,
    {
      style: [styles.container, style],
      isDismissable,
      onProfileEvent: handleProfileEvent
    }
  );
}
const styles = import_react_native.StyleSheet.create({
  container: {
    flex: 1
  },
  text: {
    fontSize: 16,
    color: "#666"
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  InlineUserProfileView
});
