// @clerk/expo/dist/native/InlineAuthView.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InlineAuthView_exports = {};
__export(InlineAuthView_exports, {
  InlineAuthView: () => InlineAuthView
});
module.exports = __toCommonJS(InlineAuthView_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_error = require("@clerk/shared/error");
var import_react = require("react");
var import_react_native = require("react-native");
var import_constants = require("../constants");
var import_singleton = require("../provider/singleton");
var import_NativeClerkAuthView = __toESM(require("../specs/NativeClerkAuthView"));
var import_token_cache = require("../token-cache");
var import_native_module = require("../utils/native-module");
function InlineAuthView({ mode = "signInOrUp", isDismissable = false }) {
  const authCompletedRef = (0, import_react.useRef)(false);
  const syncSession = (0, import_react.useCallback)(async (sessionId) => {
    var _a, _b;
    if (authCompletedRef.current) {
      return;
    }
    try {
      if ((_a = import_native_module.ClerkExpoModule) == null ? void 0 : _a.getClientToken) {
        const nativeClientToken = await import_native_module.ClerkExpoModule.getClientToken();
        if (nativeClientToken) {
          await ((_b = import_token_cache.tokenCache) == null ? void 0 : _b.saveToken(import_constants.CLERK_CLIENT_JWT_KEY, nativeClientToken));
        }
      }
      const clerkInstance = (0, import_singleton.getClerkInstance)();
      if (!clerkInstance) {
        throw new import_error.ClerkRuntimeError(
          "Clerk instance is not available. Ensure <ClerkProvider> is mounted before using <InlineAuthView>.",
          { code: "expo_inline_auth_view_clerk_instance_not_available" }
        );
      }
      const clerkRecord = clerkInstance;
      if (typeof clerkRecord.__internal_reloadInitialResources === "function") {
        await clerkRecord.__internal_reloadInitialResources();
      }
      if (typeof clerkInstance.setActive === "function") {
        await clerkInstance.setActive({ session: sessionId });
      }
      authCompletedRef.current = true;
    } catch (err) {
      if (false) {
        console.error("[InlineAuthView] Failed to sync session:", err);
      }
    }
  }, []);
  const handleAuthEvent = (0, import_react.useCallback)(
    async (event) => {
      const { type, data: rawData } = event.nativeEvent;
      if (false) {
        console.log("[InlineAuthView] onAuthEvent:", type, rawData);
      }
      const data = typeof rawData === "string" ? JSON.parse(rawData) : rawData;
      if (type === "signInCompleted" || type === "signUpCompleted") {
        const sessionId = data == null ? void 0 : data.sessionId;
        if (sessionId) {
          await syncSession(sessionId);
        } else if (false) {
          console.warn("[InlineAuthView] Auth event received but no sessionId in data:", data);
        }
      }
    },
    [syncSession]
  );
  if (!import_native_module.isNativeSupported || !import_NativeClerkAuthView.default) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_native.View, { style: [styles.container, styles.fallback], children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_native.Text, { style: styles.text, children: !import_native_module.isNativeSupported ? "Native InlineAuthView is only available on iOS and Android" : 'Native InlineAuthView requires the @clerk/expo plugin. Add "@clerk/expo" to your app.json plugins array.' }) });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_NativeClerkAuthView.default,
    {
      style: styles.container,
      mode,
      isDismissable,
      onAuthEvent: handleAuthEvent
    }
  );
}
const styles = import_react_native.StyleSheet.create({
  container: {
    flex: 1
  },
  fallback: {
    justifyContent: "center",
    alignItems: "center"
  },
  text: {
    fontSize: 16,
    color: "#666"
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  InlineAuthView
});
