// @clerk/expo/dist/native/UserButton.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var UserButton_exports = {};
__export(UserButton_exports, {
  UserButton: () => UserButton
});
module.exports = __toCommonJS(UserButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("@clerk/react");
var import_react2 = require("react");
var import_react_native = require("react-native");
var import_constants = require("../constants");
var import_token_cache = require("../token-cache");
var import_native_module = require("../utils/native-module");
function getInitials(user) {
  var _a;
  if (user == null ? void 0 : user.firstName) {
    const first = user.firstName.charAt(0).toUpperCase();
    const last = ((_a = user.lastName) == null ? void 0 : _a.charAt(0).toUpperCase()) || "";
    return first + last;
  }
  return "U";
}
function UserButton(_props) {
  var _a, _b, _c, _d;
  const [nativeUser, setNativeUser] = (0, import_react2.useState)(null);
  const presentingRef = (0, import_react2.useRef)(false);
  const clerk = (0, import_react.useClerk)();
  const { user: clerkUser } = (0, import_react.useUser)();
  (0, import_react2.useEffect)(() => {
    const fetchUser = async () => {
      var _a2, _b2;
      if (!import_native_module.isNativeSupported || !((_a2 = import_native_module.ClerkExpoModule) == null ? void 0 : _a2.getSession)) {
        return;
      }
      try {
        const result = await import_native_module.ClerkExpoModule.getSession();
        const hasSession = !!((result == null ? void 0 : result.sessionId) || ((_b2 = result == null ? void 0 : result.session) == null ? void 0 : _b2.id));
        if (hasSession && (result == null ? void 0 : result.user)) {
          setNativeUser(result.user);
        } else {
          setNativeUser(null);
        }
      } catch (err) {
        if (false) {
          console.error("[UserButton] Error fetching user:", err);
        }
      }
    };
    void fetchUser();
  }, [clerkUser == null ? void 0 : clerkUser.id]);
  const user = nativeUser != null ? nativeUser : clerkUser ? {
    id: clerkUser.id,
    firstName: (_a = clerkUser.firstName) != null ? _a : void 0,
    lastName: (_b = clerkUser.lastName) != null ? _b : void 0,
    imageUrl: (_c = clerkUser.imageUrl) != null ? _c : void 0,
    primaryEmailAddress: (_d = clerkUser.primaryEmailAddress) == null ? void 0 : _d.emailAddress
  } : null;
  const handlePress = async () => {
    var _a2, _b2, _c2, _d2, _e, _f, _g, _h, _i, _j, _k, _l;
    if (presentingRef.current) {
      return;
    }
    if (!import_native_module.isNativeSupported || !((_a2 = import_native_module.ClerkExpoModule) == null ? void 0 : _a2.presentUserProfile)) {
      return;
    }
    presentingRef.current = true;
    try {
      let hadNativeSessionBefore = false;
      if (clerkUser && ((_b2 = import_native_module.ClerkExpoModule) == null ? void 0 : _b2.getSession) && ((_c2 = import_native_module.ClerkExpoModule) == null ? void 0 : _c2.configure)) {
        const preCheck = await import_native_module.ClerkExpoModule.getSession();
        hadNativeSessionBefore = !!((preCheck == null ? void 0 : preCheck.sessionId) || ((_d2 = preCheck == null ? void 0 : preCheck.session) == null ? void 0 : _d2.id));
        if (!hadNativeSessionBefore) {
          const bearerToken = (_f = await ((_e = import_token_cache.tokenCache) == null ? void 0 : _e.getToken(import_constants.CLERK_CLIENT_JWT_KEY))) != null ? _f : null;
          if (bearerToken) {
            await import_native_module.ClerkExpoModule.configure(clerk.publishableKey, bearerToken);
            const postConfigure = await import_native_module.ClerkExpoModule.getSession();
            hadNativeSessionBefore = !!((postConfigure == null ? void 0 : postConfigure.sessionId) || ((_g = postConfigure == null ? void 0 : postConfigure.session) == null ? void 0 : _g.id));
          }
        }
      }
      await import_native_module.ClerkExpoModule.presentUserProfile({
        dismissable: true
      });
      const sessionCheck = await ((_i = (_h = import_native_module.ClerkExpoModule).getSession) == null ? void 0 : _i.call(_h));
      const hasNativeSession = !!((sessionCheck == null ? void 0 : sessionCheck.sessionId) || ((_j = sessionCheck == null ? void 0 : sessionCheck.session) == null ? void 0 : _j.id));
      if (!hasNativeSession && hadNativeSessionBefore) {
        setNativeUser(null);
        try {
          await ((_l = (_k = import_native_module.ClerkExpoModule).signOut) == null ? void 0 : _l.call(_k));
        } catch (e) {
          if (false) {
            console.warn("[UserButton] Native signOut error (may already be signed out):", e);
          }
        }
        if (clerk == null ? void 0 : clerk.signOut) {
          try {
            await clerk.signOut();
          } catch (e) {
            if (false) {
              console.warn("[UserButton] JS SDK signOut error:", e);
            }
          }
        }
      }
    } catch (error) {
      if (false) {
        console.error("[UserButton] presentUserProfile failed:", error);
      }
    } finally {
      presentingRef.current = false;
    }
  };
  if (!import_native_module.isNativeSupported || !import_native_module.ClerkExpoModule) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_native.View, { style: styles.button, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_native.Text, { style: styles.text, children: "?" }) });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_react_native.TouchableOpacity,
    {
      onPress: () => void handlePress(),
      style: styles.button,
      children: (user == null ? void 0 : user.imageUrl) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_native.Image,
        {
          source: { uri: user.imageUrl },
          style: styles.avatarImage
        }
      ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_native.View, { style: styles.avatar, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_native.Text, { style: styles.avatarText, children: getInitials(user) }) })
    }
  );
}
const styles = import_react_native.StyleSheet.create({
  button: {
    width: "100%",
    height: "100%",
    overflow: "hidden"
  },
  avatar: {
    flex: 1,
    backgroundColor: "#6366f1",
    justifyContent: "center",
    alignItems: "center"
  },
  avatarImage: {
    width: "100%",
    height: "100%"
  },
  avatarText: {
    color: "white",
    fontSize: 14,
    fontWeight: "600"
  },
  text: {
    fontSize: 14,
    color: "#666"
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  UserButton
});
