// @clerk/expo/dist/web/uiComponents.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var uiComponents_exports = {};
__export(uiComponents_exports, {
  CreateOrganization: () => CreateOrganization,
  GoogleOneTap: () => GoogleOneTap,
  OrganizationList: () => OrganizationList,
  OrganizationProfile: () => OrganizationProfile,
  OrganizationSwitcher: () => OrganizationSwitcher,
  PricingTable: () => PricingTable,
  SignIn: () => SignIn,
  SignInButton: () => SignInButton,
  SignInWithMetamaskButton: () => SignInWithMetamaskButton,
  SignOutButton: () => SignOutButton,
  SignUp: () => SignUp,
  SignUpButton: () => SignUpButton,
  UNSAFE_PortalProvider: () => UNSAFE_PortalProvider,
  UserAvatar: () => UserAvatar,
  UserButton: () => UserButton,
  UserProfile: () => UserProfile,
  Waitlist: () => Waitlist
});
module.exports = __toCommonJS(uiComponents_exports);
var import_react = require("@clerk/react");
var import_react_native = require("react-native");
var import_errors = require("../utils/errors");
const ErrorComponent = (componentName) => {
  import_errors.errorThrower.throw(`<${componentName}/> component is not supported in native environments.`);
};
function WrapComponent(component) {
  const commonentName = component.displayName.replace("withClerk(", "").replace(")", "");
  return import_react_native.Platform.select({
    // @ts-expect-error - This is just to make the function signature match the original
    native: () => () => ErrorComponent(commonentName),
    default: () => component
  })();
}
const SignIn = WrapComponent(import_react.SignIn);
const SignUp = WrapComponent(import_react.SignUp);
const UserAvatar = WrapComponent(import_react.UserAvatar);
const UserButton = WrapComponent(import_react.UserButton);
const UserProfile = WrapComponent(import_react.UserProfile);
const OrganizationProfile = WrapComponent(import_react.OrganizationProfile);
const OrganizationSwitcher = WrapComponent(import_react.OrganizationSwitcher);
const OrganizationList = WrapComponent(import_react.OrganizationList);
const CreateOrganization = WrapComponent(import_react.CreateOrganization);
const SignInButton = WrapComponent(import_react.SignInButton);
const SignUpButton = WrapComponent(import_react.SignUpButton);
const SignOutButton = WrapComponent(import_react.SignOutButton);
const SignInWithMetamaskButton = WrapComponent(import_react.SignInWithMetamaskButton);
const GoogleOneTap = WrapComponent(import_react.GoogleOneTap);
const Waitlist = WrapComponent(import_react.Waitlist);
const PricingTable = WrapComponent(import_react.PricingTable);
const UNSAFE_PortalProvider = WrapComponent(import_react.UNSAFE_PortalProvider);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CreateOrganization,
  GoogleOneTap,
  OrganizationList,
  OrganizationProfile,
  OrganizationSwitcher,
  PricingTable,
  SignIn,
  SignInButton,
  SignInWithMetamaskButton,
  SignOutButton,
  SignUp,
  SignUpButton,
  UNSAFE_PortalProvider,
  UserAvatar,
  UserButton,
  UserProfile,
  Waitlist
});
