// @clerk/expo/dist/hooks/useSignInWithGoogle.shared.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useSignInWithGoogle_shared_exports = {};
__export(useSignInWithGoogle_shared_exports, {
  createUseSignInWithGoogle: () => createUseSignInWithGoogle,
  executeGoogleAuthenticationFlow: () => executeGoogleAuthenticationFlow
});
module.exports = __toCommonJS(useSignInWithGoogle_shared_exports);
var import_react = require("@clerk/react");
var import_error = require("@clerk/shared/error");
var import_google_one_tap = require("../google-one-tap");
var import_errors = require("../utils/errors");
async function getGoogleClientIds() {
  var _a, _b, _c, _d;
  let webClientId;
  let iosClientId;
  try {
    const ConstantsModule = await import("expo-constants");
    const Constants = ConstantsModule.default;
    webClientId = ((_b = (_a = Constants == null ? void 0 : Constants.expoConfig) == null ? void 0 : _a.extra) == null ? void 0 : _b.EXPO_PUBLIC_CLERK_GOOGLE_WEB_CLIENT_ID) || process.env.EXPO_PUBLIC_CLERK_GOOGLE_WEB_CLIENT_ID;
    iosClientId = ((_d = (_c = Constants == null ? void 0 : Constants.expoConfig) == null ? void 0 : _c.extra) == null ? void 0 : _d.EXPO_PUBLIC_CLERK_GOOGLE_IOS_CLIENT_ID) || process.env.EXPO_PUBLIC_CLERK_GOOGLE_IOS_CLIENT_ID;
  } catch {
    webClientId = process.env.EXPO_PUBLIC_CLERK_GOOGLE_WEB_CLIENT_ID;
    iosClientId = process.env.EXPO_PUBLIC_CLERK_GOOGLE_IOS_CLIENT_ID;
  }
  return { webClientId, iosClientId };
}
function createUseSignInWithGoogle(platformConfig) {
  return function useSignInWithGoogle() {
    const clerk = (0, import_react.useClerk)();
    async function startGoogleAuthenticationFlow(startGoogleAuthenticationFlowParams) {
      const { client, loaded, setActive } = clerk;
      if (!loaded || !client) {
        return {
          createdSessionId: null,
          setActive
        };
      }
      const { webClientId, iosClientId } = await getGoogleClientIds();
      if (!webClientId) {
        return import_errors.errorThrower.throw(
          "Google Sign-In credentials not found. Please set EXPO_PUBLIC_CLERK_GOOGLE_WEB_CLIENT_ID in your .env file."
        );
      }
      if (platformConfig.requiresIosClientId && !iosClientId) {
        return import_errors.errorThrower.throw(
          "Google Sign-In credentials not found. Please set EXPO_PUBLIC_CLERK_GOOGLE_IOS_CLIENT_ID in your .env file."
        );
      }
      return executeGoogleAuthenticationFlow(
        { client, setActive },
        { webClientId, iosClientId },
        startGoogleAuthenticationFlowParams
      );
    }
    return {
      startGoogleAuthenticationFlow
    };
  };
}
async function executeGoogleAuthenticationFlow(context, clientIds, params) {
  var _a;
  const { client, setActive } = context;
  const { signIn, signUp } = client;
  import_google_one_tap.ClerkGoogleOneTapSignIn.configure(clientIds);
  const { randomUUID } = await import("expo-crypto");
  const nonce = randomUUID();
  try {
    const response = await import_google_one_tap.ClerkGoogleOneTapSignIn.presentExplicitSignIn({
      nonce
    });
    if (!(0, import_google_one_tap.isSuccessResponse)(response)) {
      return {
        createdSessionId: null,
        setActive,
        signIn,
        signUp
      };
    }
    const { idToken } = response.data;
    try {
      await signIn.create({
        strategy: "google_one_tap",
        token: idToken
      });
      const userNeedsToBeCreated = signIn.firstFactorVerification.status === "transferable";
      if (userNeedsToBeCreated) {
        await signUp.create({
          transfer: true,
          unsafeMetadata: params == null ? void 0 : params.unsafeMetadata
        });
        return {
          createdSessionId: signUp.createdSessionId,
          setActive,
          signIn,
          signUp
        };
      }
      return {
        createdSessionId: signIn.createdSessionId,
        setActive,
        signIn,
        signUp
      };
    } catch (signInError) {
      if ((0, import_error.isClerkAPIResponseError)(signInError) && ((_a = signInError.errors) == null ? void 0 : _a.some((err) => err.code === "external_account_not_found"))) {
        await signUp.create({
          strategy: "google_one_tap",
          token: idToken,
          unsafeMetadata: params == null ? void 0 : params.unsafeMetadata
        });
        return {
          createdSessionId: signUp.createdSessionId,
          setActive,
          signIn,
          signUp
        };
      }
      throw signInError;
    }
  } catch (error) {
    if ((0, import_google_one_tap.isErrorWithCode)(error) && error.code === "SIGN_IN_CANCELLED") {
      return {
        createdSessionId: null,
        setActive,
        signIn,
        signUp
      };
    }
    throw error;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createUseSignInWithGoogle,
  executeGoogleAuthenticationFlow
});
