// @clerk/expo/dist/google-one-tap/ClerkGoogleOneTapSignIn.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ClerkGoogleOneTapSignIn_exports = {};
__export(ClerkGoogleOneTapSignIn_exports, {
  ClerkGoogleOneTapSignIn: () => ClerkGoogleOneTapSignIn,
  isCancelledResponse: () => isCancelledResponse,
  isErrorWithCode: () => isErrorWithCode,
  isNoSavedCredentialFoundResponse: () => isNoSavedCredentialFoundResponse,
  isSuccessResponse: () => isSuccessResponse
});
module.exports = __toCommonJS(ClerkGoogleOneTapSignIn_exports);
var import_NativeClerkGoogleSignIn = __toESM(require("../specs/NativeClerkGoogleSignIn"));
function getNativeModule() {
  if (!import_NativeClerkGoogleSignIn.default) {
    throw new Error(
      "ClerkGoogleSignIn native module is not available. Ensure the @clerk/expo plugin is added to your app.json and you have run a development build."
    );
  }
  return import_NativeClerkGoogleSignIn.default;
}
function isCancelledResponse(response) {
  return response.type === "cancelled";
}
function isNoSavedCredentialFoundResponse(response) {
  return response.type === "noSavedCredentialFound";
}
function isSuccessResponse(response) {
  return response.type === "success";
}
function isErrorWithCode(error) {
  return error !== null && typeof error === "object" && "code" in error && typeof error.code === "string" && "message" in error && typeof error.message === "string";
}
const ClerkGoogleOneTapSignIn = {
  /**
   * Configure Google Sign-In. Must be called before any sign-in methods.
   *
   * @param params - Configuration parameters
   * @param params.webClientId - The web client ID from Google Cloud Console (required)
   * @param params.hostedDomain - Optional domain restriction
   * @param params.autoSelectEnabled - Auto-select for single credential (default: false)
   */
  configure(params) {
    getNativeModule().configure(params);
  },
  /**
   * Attempt to sign in with saved credentials (One Tap).
   *
   * This method will show the One Tap UI if there are saved credentials,
   * or return a "noSavedCredentialFound" response if there are none.
   *
   * @param params - Sign-in parameters
   * @param params.nonce - Cryptographic nonce for replay protection
   * @param params.filterByAuthorizedAccounts - Only show previously authorized accounts (default: true)
   *
   * @returns Promise resolving to OneTapResponse
   */
  async signIn(params) {
    try {
      return await getNativeModule().signIn(params != null ? params : null);
    } catch (error) {
      if (isErrorWithCode(error)) {
        if (error.code === "SIGN_IN_CANCELLED") {
          return { type: "cancelled", data: null };
        }
        if (error.code === "NO_SAVED_CREDENTIAL_FOUND") {
          return { type: "noSavedCredentialFound", data: null };
        }
      }
      throw error;
    }
  },
  /**
   * Create a new account (shows all Google accounts).
   *
   * This method shows the account picker with all available Google accounts,
   * not just previously authorized ones.
   *
   * @param params - Create account parameters
   * @param params.nonce - Cryptographic nonce for replay protection
   *
   * @returns Promise resolving to OneTapResponse
   */
  async createAccount(params) {
    try {
      return await getNativeModule().createAccount(params != null ? params : null);
    } catch (error) {
      if (isErrorWithCode(error)) {
        if (error.code === "SIGN_IN_CANCELLED") {
          return { type: "cancelled", data: null };
        }
        if (error.code === "NO_SAVED_CREDENTIAL_FOUND") {
          return { type: "noSavedCredentialFound", data: null };
        }
      }
      throw error;
    }
  },
  /**
   * Present explicit sign-in UI (Google Sign-In button flow).
   *
   * This shows the full Google Sign-In UI, similar to clicking a
   * "Sign in with Google" button.
   *
   * @param params - Explicit sign-in parameters
   * @param params.nonce - Cryptographic nonce for replay protection
   *
   * @returns Promise resolving to OneTapResponse
   */
  async presentExplicitSignIn(params) {
    try {
      return await getNativeModule().presentExplicitSignIn(params != null ? params : null);
    } catch (error) {
      if (isErrorWithCode(error)) {
        if (error.code === "SIGN_IN_CANCELLED") {
          return { type: "cancelled", data: null };
        }
      }
      throw error;
    }
  },
  /**
   * Sign out and clear credential state.
   *
   * This disables automatic sign-in until the user signs in again.
   */
  async signOut() {
    await getNativeModule().signOut();
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ClerkGoogleOneTapSignIn,
  isCancelledResponse,
  isErrorWithCode,
  isNoSavedCredentialFoundResponse,
  isSuccessResponse
});
