// @clerk/expo/dist/provider/ClerkProvider.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ClerkProvider_exports = {};
__export(ClerkProvider_exports, {
  ClerkProvider: () => ClerkProvider
});
module.exports = __toCommonJS(ClerkProvider_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_polyfills = require("../polyfills");
var import_react = require("@clerk/react");
var import_internal = require("@clerk/react/internal");
var import_react2 = require("react");
var import_react_native = require("react-native");
var import_constants = require("../constants");
var import_useNativeAuthEvents = require("../hooks/useNativeAuthEvents");
var import_NativeClerkModule = __toESM(require("../specs/NativeClerkModule"));
var import_token_cache = require("../token-cache");
var import_runtime = require("../utils/runtime");
var import_singleton = require("./singleton");
const SDK_METADATA = {
  name: "@clerk/expo",
  version: "3.2.9"
};
function NativeSessionSync({
  publishableKey,
  tokenCache
}) {
  const { isSignedIn, isLoaded } = (0, import_react.useAuth)();
  const hasSyncedRef = (0, import_react2.useRef)(false);
  const effectiveTokenCache = tokenCache != null ? tokenCache : import_token_cache.tokenCache;
  (0, import_react2.useEffect)(() => {
    if (!isSignedIn) {
      hasSyncedRef.current = false;
      if (isLoaded) {
        const ClerkExpo = import_NativeClerkModule.default;
        if (ClerkExpo == null ? void 0 : ClerkExpo.signOut) {
          void ClerkExpo.signOut().catch((error) => {
            if (false) {
              console.warn("[NativeSessionSync] Failed to clear native session:", error);
            }
          });
        }
      }
      return;
    }
    if (hasSyncedRef.current) {
      return;
    }
    const syncToNative = async () => {
      var _a, _b;
      try {
        const ClerkExpo = import_NativeClerkModule.default;
        if (!(ClerkExpo == null ? void 0 : ClerkExpo.configure) || !(ClerkExpo == null ? void 0 : ClerkExpo.getSession)) {
          return;
        }
        const nativeSession = await ClerkExpo.getSession();
        const hasNativeSession = !!((nativeSession == null ? void 0 : nativeSession.sessionId) || ((_a = nativeSession == null ? void 0 : nativeSession.session) == null ? void 0 : _a.id));
        if (hasNativeSession) {
          hasSyncedRef.current = true;
          return;
        }
        const bearerToken = (_b = await (effectiveTokenCache == null ? void 0 : effectiveTokenCache.getToken(import_constants.CLERK_CLIENT_JWT_KEY))) != null ? _b : null;
        if (bearerToken) {
          await ClerkExpo.configure(publishableKey, bearerToken);
          hasSyncedRef.current = true;
        }
      } catch (error) {
        if (false) {
          console.warn("[NativeSessionSync] Failed to sync JS session to native:", error);
        }
      }
    };
    void syncToNative();
  }, [isSignedIn, isLoaded, publishableKey, effectiveTokenCache]);
  return null;
}
function ClerkProvider(props) {
  const {
    children,
    tokenCache,
    publishableKey,
    proxyUrl,
    domain,
    __experimental_passkeys,
    experimental,
    __experimental_resourceCache,
    ...rest
  } = props;
  const pk = publishableKey;
  const pendingNativeSessionRef = (0, import_react2.useRef)(null);
  const initStartedRef = (0, import_react2.useRef)(false);
  const sessionSyncedRef = (0, import_react2.useRef)(false);
  (0, import_react2.useEffect)(() => {
    pendingNativeSessionRef.current = null;
    initStartedRef.current = false;
    sessionSyncedRef.current = false;
  }, [pk]);
  const clerkInstance = (0, import_runtime.isNative)() ? (0, import_singleton.getClerkInstance)({
    publishableKey: pk,
    tokenCache,
    proxyUrl,
    domain,
    __experimental_passkeys,
    __experimental_resourceCache
  }) : null;
  const isMountedRef = (0, import_react2.useRef)(true);
  (0, import_react2.useEffect)(() => {
    isMountedRef.current = true;
    if ((import_react_native.Platform.OS === "ios" || import_react_native.Platform.OS === "android") && pk && !initStartedRef.current) {
      initStartedRef.current = true;
      const configureNativeClerk = async () => {
        var _a, _b, _c, _d, _e, _f;
        try {
          const ClerkExpo = import_NativeClerkModule.default;
          if (ClerkExpo == null ? void 0 : ClerkExpo.configure) {
            const effectiveTokenCache = tokenCache != null ? tokenCache : import_token_cache.tokenCache;
            let bearerToken = null;
            try {
              bearerToken = (_a = await (effectiveTokenCache == null ? void 0 : effectiveTokenCache.getToken(import_constants.CLERK_CLIENT_JWT_KEY))) != null ? _a : null;
            } catch (e) {
              if (false) {
                console.warn("[ClerkProvider] Token cache read failed:", e);
              }
            }
            await ClerkExpo.configure(pk, bearerToken);
            if (!isMountedRef.current) {
              return;
            }
            const MAX_WAIT_MS = 3e3;
            const POLL_INTERVAL_MS = 100;
            let sessionId = null;
            for (let elapsed = 0; elapsed < MAX_WAIT_MS; elapsed += POLL_INTERVAL_MS) {
              if (!isMountedRef.current) {
                return;
              }
              if (ClerkExpo == null ? void 0 : ClerkExpo.getSession) {
                const nativeSession = await ClerkExpo.getSession();
                sessionId = (_d = (_c = nativeSession == null ? void 0 : nativeSession.sessionId) != null ? _c : (_b = nativeSession == null ? void 0 : nativeSession.session) == null ? void 0 : _b.id) != null ? _d : null;
                if (sessionId) {
                  break;
                }
              }
              await new Promise((resolve) => setTimeout(resolve, POLL_INTERVAL_MS));
            }
            if (!isMountedRef.current) {
              return;
            }
            if (sessionId && clerkInstance) {
              pendingNativeSessionRef.current = sessionId;
              const clerkAny = clerkInstance;
              const waitForLoad = () => {
                return new Promise((resolve) => {
                  if (clerkAny.loaded) {
                    resolve();
                  } else if (typeof clerkAny.addOnLoaded === "function") {
                    clerkAny.addOnLoaded(() => resolve());
                  } else {
                    if (false) {
                      console.warn("[ClerkProvider] Clerk instance has no loaded property or addOnLoaded method");
                    }
                    resolve();
                  }
                });
              };
              await waitForLoad();
              if (!isMountedRef.current) {
                return;
              }
              if (!sessionSyncedRef.current && typeof clerkInstance.setActive === "function") {
                sessionSyncedRef.current = true;
                const pendingSession = pendingNativeSessionRef.current;
                const sessionInClient = (_f = (_e = clerkInstance.client) == null ? void 0 : _e.sessions) == null ? void 0 : _f.some(
                  (s) => s.id === pendingSession
                );
                if (!sessionInClient && typeof clerkAny.__internal_reloadInitialResources === "function") {
                  await clerkAny.__internal_reloadInitialResources();
                }
                try {
                  await clerkInstance.setActive({ session: pendingSession });
                } catch (err) {
                  if (false) {
                    console.error(`[ClerkProvider] Failed to sync native session:`, err);
                  }
                }
              }
            }
          }
        } catch (error) {
          const isNativeModuleNotFound = error instanceof Error && (error.message.includes("Cannot find native module") || error.message.includes("TurboModuleRegistry.getEnforcing(...): 'ClerkExpo'"));
          if (isNativeModuleNotFound) {
            if (false) {
              console.debug(
                `[ClerkProvider] Native Clerk module not available. To enable native features, add "@clerk/expo" to your app.json plugins array.`
              );
            }
          } else if (false) {
            console.error(`[ClerkProvider] Failed to configure Clerk ${import_react_native.Platform.OS}:`, error);
          }
        }
      };
      void configureNativeClerk();
    }
    return () => {
      isMountedRef.current = false;
    };
  }, [pk, clerkInstance]);
  const { nativeAuthState } = (0, import_useNativeAuthEvents.useNativeAuthEvents)();
  (0, import_react2.useEffect)(() => {
    if (!nativeAuthState || !clerkInstance) {
      return;
    }
    const syncNativeAuthToJs = async () => {
      var _a, _b;
      try {
        if (nativeAuthState.type === "signedIn" && nativeAuthState.sessionId && clerkInstance.setActive) {
          const ClerkExpo = import_NativeClerkModule.default;
          if (ClerkExpo == null ? void 0 : ClerkExpo.getClientToken) {
            const nativeClientToken = await ClerkExpo.getClientToken();
            if (nativeClientToken) {
              const effectiveTokenCache = tokenCache != null ? tokenCache : import_token_cache.tokenCache;
              await (effectiveTokenCache == null ? void 0 : effectiveTokenCache.saveToken(import_constants.CLERK_CLIENT_JWT_KEY, nativeClientToken));
            }
          }
          const sessionInClient = (_b = (_a = clerkInstance.client) == null ? void 0 : _a.sessions) == null ? void 0 : _b.some(
            (s) => s.id === nativeAuthState.sessionId
          );
          if (!sessionInClient) {
            const clerkAny = clerkInstance;
            if (typeof clerkAny.__internal_reloadInitialResources === "function") {
              await clerkAny.__internal_reloadInitialResources();
            }
            if (!isMountedRef.current) {
              return;
            }
          }
          if (!isMountedRef.current) {
            return;
          }
          await clerkInstance.setActive({ session: nativeAuthState.sessionId });
        } else if (nativeAuthState.type === "signedOut" && clerkInstance.signOut) {
          if (!isMountedRef.current) {
            return;
          }
          await clerkInstance.signOut();
        }
      } catch (error) {
        if (false) {
          console.error(`[ClerkProvider] Failed to sync native auth state:`, error);
        }
      }
    };
    void syncNativeAuthToJs();
  }, [nativeAuthState, clerkInstance]);
  if ((0, import_runtime.isWeb)()) {
    try {
      const WebBrowser = require("expo-web-browser");
      WebBrowser.maybeCompleteAuthSession();
    } catch (e) {
      if (false) {
        console.warn("[ClerkProvider] expo-web-browser not available, OAuth/SSO on web will not work:", e);
      }
    }
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_internal.InternalClerkProvider,
    {
      ...rest,
      publishableKey: pk,
      proxyUrl,
      domain,
      sdkMetadata: SDK_METADATA,
      Clerk: clerkInstance,
      standardBrowser: !(0, import_runtime.isNative)(),
      experimental: {
        ...experimental,
        // force the rethrowOfflineNetworkErrors flag to true if the asyncStorage is provided
        rethrowOfflineNetworkErrors: !!__experimental_resourceCache || (experimental == null ? void 0 : experimental.rethrowOfflineNetworkErrors),
        ...(0, import_runtime.isNative)() && { runtimeEnvironment: "headless" }
      },
      children: [
        (0, import_runtime.isNative)() && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          NativeSessionSync,
          {
            publishableKey: pk,
            tokenCache
          }
        ),
        children
      ]
    },
    pk
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ClerkProvider
});
