// @clerk/expo/dist/provider/singleton/createClerkInstance.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var createClerkInstance_exports = {};
__export(createClerkInstance_exports, {
  createClerkInstance: () => createClerkInstance
});
module.exports = __toCommonJS(createClerkInstance_exports);
var import_error = require("@clerk/shared/error");
var import_react_native = require("react-native");
var import_package = __toESM(require("../../../package.json"));
var import_cache = require("../../cache");
var import_MemoryTokenCache = require("../../cache/MemoryTokenCache");
var import_constants = require("../../constants");
var import_errorThrower = require("../../errorThrower");
var import_utils = require("../../utils");
function hasOwnOption(options, key) {
  return !!options && Object.prototype.hasOwnProperty.call(options, key);
}
let __internal_clerk;
let __internal_clerkOptions;
function getUpdatedClerkOptions(currentOptions, nextOptions) {
  var _a, _b;
  const hasNextProxyUrl = hasOwnOption(nextOptions, "proxyUrl");
  const hasNextDomain = hasOwnOption(nextOptions, "domain");
  const hasKeyChanged = !!currentOptions && typeof (nextOptions == null ? void 0 : nextOptions.publishableKey) !== "undefined" && nextOptions.publishableKey !== currentOptions.publishableKey;
  const hasProxyChanged = !!currentOptions && hasNextProxyUrl && nextOptions.proxyUrl !== currentOptions.proxyUrl;
  const hasDomainChanged = !!currentOptions && hasNextDomain && nextOptions.domain !== currentOptions.domain;
  return {
    hasConfigChanged: hasKeyChanged || hasProxyChanged || hasDomainChanged,
    options: {
      publishableKey: (_b = (_a = nextOptions == null ? void 0 : nextOptions.publishableKey) != null ? _a : currentOptions == null ? void 0 : currentOptions.publishableKey) != null ? _b : "",
      proxyUrl: hasKeyChanged ? nextOptions == null ? void 0 : nextOptions.proxyUrl : hasNextProxyUrl ? nextOptions.proxyUrl : currentOptions == null ? void 0 : currentOptions.proxyUrl,
      domain: hasKeyChanged ? nextOptions == null ? void 0 : nextOptions.domain : hasNextDomain ? nextOptions.domain : currentOptions == null ? void 0 : currentOptions.domain
    }
  };
}
function createClerkInstance(ClerkClass) {
  return (options) => {
    var _a;
    const { tokenCache = import_MemoryTokenCache.MemoryTokenCache, __experimental_resourceCache: createResourceCache } = options || {};
    const {
      hasConfigChanged,
      options: { publishableKey, proxyUrl, domain }
    } = getUpdatedClerkOptions(__internal_clerkOptions, options);
    if (!__internal_clerk && !publishableKey) {
      import_errorThrower.errorThrower.throwMissingPublishableKeyError();
    }
    if (!__internal_clerk || hasConfigChanged) {
      (0, import_utils.assertValidProxyUrl)(proxyUrl);
      if (hasConfigChanged) {
        (_a = tokenCache.clearToken) == null ? void 0 : _a.call(tokenCache, import_constants.CLERK_CLIENT_JWT_KEY);
      }
      const getToken = tokenCache.getToken;
      const saveToken = tokenCache.saveToken;
      __internal_clerkOptions = { publishableKey, proxyUrl, domain };
      __internal_clerk = new ClerkClass(publishableKey, { proxyUrl, domain });
      if (import_react_native.Platform.OS === "ios" || import_react_native.Platform.OS === "android") {
        __internal_clerk.__internal_createPublicCredentials = (publicKeyCredential) => {
          var _a2, _b;
          return ((_a2 = options == null ? void 0 : options.__experimental_passkeys) == null ? void 0 : _a2.create) ? (_b = options == null ? void 0 : options.__experimental_passkeys) == null ? void 0 : _b.create(publicKeyCredential) : import_errorThrower.errorThrower.throw("create() for passkeys is missing");
        };
        __internal_clerk.__internal_getPublicCredentials = ({
          publicKeyOptions
        }) => {
          var _a2, _b;
          return ((_a2 = options == null ? void 0 : options.__experimental_passkeys) == null ? void 0 : _a2.get) ? (_b = options == null ? void 0 : options.__experimental_passkeys) == null ? void 0 : _b.get({ publicKeyOptions }) : import_errorThrower.errorThrower.throw("get() for passkeys is missing");
        };
        __internal_clerk.__internal_isWebAuthnSupported = () => {
          var _a2, _b;
          return ((_a2 = options == null ? void 0 : options.__experimental_passkeys) == null ? void 0 : _a2.isSupported) ? (_b = options == null ? void 0 : options.__experimental_passkeys) == null ? void 0 : _b.isSupported() : import_errorThrower.errorThrower.throw("isSupported() for passkeys is missing");
        };
        __internal_clerk.__internal_isWebAuthnAutofillSupported = () => {
          var _a2, _b;
          return ((_a2 = options == null ? void 0 : options.__experimental_passkeys) == null ? void 0 : _a2.isAutoFillSupported) ? (_b = options == null ? void 0 : options.__experimental_passkeys) == null ? void 0 : _b.isAutoFillSupported() : import_errorThrower.errorThrower.throw("isSupported() for passkeys is missing");
        };
        __internal_clerk.__internal_isWebAuthnPlatformAuthenticatorSupported = () => {
          return Promise.resolve(true);
        };
        const isClerkNetworkError = (err) => (0, import_error.isClerkRuntimeError)(err) && err.code === "network_error";
        if (createResourceCache) {
          const retryInitilizeResourcesFromFAPI = async () => {
            try {
              await (__internal_clerk == null ? void 0 : __internal_clerk.__internal_reloadInitialResources());
            } catch (err) {
              if (isClerkNetworkError(err) || !(0, import_error.is4xxError)(err)) {
                const timeout = isClerkNetworkError(err) ? 2e3 : 1e4;
                setTimeout(() => void retryInitilizeResourcesFromFAPI(), timeout);
              }
            }
          };
          import_cache.EnvironmentResourceCache.init({ publishableKey, storage: createResourceCache });
          import_cache.ClientResourceCache.init({ publishableKey, storage: createResourceCache });
          import_cache.SessionJWTCache.init({ publishableKey, storage: createResourceCache });
          __internal_clerk.addListener(({ client }) => {
            var _a2;
            const environment = __internal_clerk == null ? void 0 : __internal_clerk.__internal_environment;
            if (environment) {
              void import_cache.EnvironmentResourceCache.save(environment.__internal_toSnapshot());
            }
            if (client) {
              void import_cache.ClientResourceCache.save(client.__internal_toSnapshot());
              if (client.lastActiveSessionId) {
                const currentSession = client.signedInSessions.find((s) => s.id === client.lastActiveSessionId);
                const token = (_a2 = currentSession == null ? void 0 : currentSession.lastActiveToken) == null ? void 0 : _a2.getRawString();
                if (token) {
                  void import_cache.SessionJWTCache.save(token);
                }
              } else {
                void import_cache.SessionJWTCache.remove();
              }
            }
          });
          __internal_clerk.__internal_getCachedResources = async () => {
            let environment = await import_cache.EnvironmentResourceCache.load();
            let client = await import_cache.ClientResourceCache.load();
            if (!environment || !client) {
              environment = import_cache.DUMMY_CLERK_ENVIRONMENT_RESOURCE;
              client = import_cache.DUMMY_CLERK_CLIENT_RESOURCE;
              setTimeout(() => void retryInitilizeResourcesFromFAPI(), 3e3);
            }
            return { client, environment };
          };
        }
      }
      __internal_clerk.__internal_onBeforeRequest(async (requestInit) => {
        var _a2;
        requestInit.credentials = "omit";
        (_a2 = requestInit.url) == null ? void 0 : _a2.searchParams.append("_is_native", "1");
        const jwt = await getToken(import_constants.CLERK_CLIENT_JWT_KEY);
        requestInit.headers.set("authorization", jwt || "");
        if ((0, import_utils.isNative)()) {
          requestInit.headers.set("x-mobile", "1");
          requestInit.headers.set("x-expo-sdk-version", import_package.default.version);
        }
      });
      let nativeApiErrorShown = false;
      __internal_clerk.__internal_onAfterResponse(async (_, response) => {
        const authHeader = response.headers.get("authorization");
        if (authHeader) {
          await saveToken(import_constants.CLERK_CLIENT_JWT_KEY, authHeader);
        }
        if (false) {
          console.error(
            "The Native API is disabled for this instance.\nGo to Clerk Dashboard > Configure > Native applications to enable it.\nOr, navigate here: https://dashboard.clerk.com/last-active?path=native-applications"
          );
          nativeApiErrorShown = true;
        }
      });
    }
    return __internal_clerk;
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createClerkInstance
});
